toolkit/tools/scheduler/schedulerutils/graphbuildstate.go

```diff
@@ -17,31 +18,52 @@ type nodeState struct {
 	available bool
 	cached    bool
 	usedDelta bool
+	freshness int
 }
 
+const (
+	// Negative values indicate that the node was rebuilt due to missing files and should gain the maximum configured freshness value.
+	NodeFreshnessRebuildRequired int = -1
+	// The highest possible freshness value, used to force unbounded cascading rebuilds.
+	NodeFreshnessAbsoluteMax int = math.MaxInt
+)
+
 // GraphBuildState represents the build state of a graph.
 type GraphBuildState struct {
 	activeBuilds     map[int64]*BuildRequest
 	nodeToState      map[*pkggraph.PkgNode]*nodeState
+	maxFreshness     int
 	failures         []*BuildResult
 	reservedFiles    map[string]bool
 	conflictingRPMs  map[string]bool
 	conflictingSRPMs map[string]bool
 }
 
 // NewGraphBuildState returns a new GraphBuildState.
-// - reservedFiles is a list of reserved files which should NOT be rebuilt. Any files that ARE rebuilt will be recorded.
-func NewGraphBuildState(reservedFiles []string) (g *GraphBuildState) {
+//   - reservedFiles is a list of reserved files which should NOT be rebuilt. Any files that ARE rebuilt will be recorded.
+//   - maxFreshness is how fresh a newly rebuilt node is. Each dependant node will have a freshness of 'n-1', etc. until
+//     '0' where the subsequent nodes will no longer be rebuilt. 'maxFreshness < 0' will cause unbounded cascading rebuilds,
+//     while 'maxFreshness = 0' will cause no cascading rebuilds.
+func NewGraphBuildState(reservedFiles []string, maxFreshness int) (g *GraphBuildState) {
 	filesMap := make(map[string]bool)
 	for _, file := range reservedFiles {
 		filesMap[file] = true
 	}
+
+	//Use max int to represent infinity. This will cause unbounded cascading rebuilds so long as we build less that 2^31
+	// or more likely 2^63 packages (i.e., ~2.3 billion packages)
```

Comment:
I think we can be a bit more concise with our comments...;) Also, in this particular place we're checking for negative numbers, so this might be a bit confusing.